superduperdb/ext/llm/base.py

```diff
@@ -0,0 +1,317 @@
+import abc
+import asyncio
+import dataclasses as dc
+import inspect
+import typing
+from functools import reduce, wraps
+from logging import WARNING, getLogger
+from typing import Any, Callable, List, Optional, Union
+
+from superduperdb import logging
+from superduperdb.components.component import Component
+from superduperdb.components.model import _Predictor
+from superduperdb.ext.utils import format_prompt
+
+if typing.TYPE_CHECKING:
+    from superduperdb.base.datalayer import Datalayer
+
+# Disable httpx info level logging
+getLogger("httpx").setLevel(WARNING)
+
+
+def ensure_initialized(func):
+    """Decorator to ensure that the model is initialized before calling the function"""
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        if not hasattr(self, "_is_initialized") or not self._is_initialized:
+            model_message = f"{self.__class__.__name__} : {self.identifier}"
+            logging.info(f"Initializing {model_message}")
+            self.init()
+            self._is_initialized = True
+            logging.info(f"Initialized  {model_message} successfully")
+        return func(self, *args, **kwargs)
+
+    return wrapper
+
+
+@dc.dataclass
+class _BaseLLM(Component, _Predictor, metaclass=abc.ABCMeta):
+    """
+    :param prompt_template: The template to use for the prompt.
+    :param prompt_func: The function to use for the prompt.
+    :param max_batch_size: The maximum batch size to use for batch generation.
+    :param inference_kwargs: Parameters used during inference.
+    """
+
+    prompt_template: str = "{input}"
+    prompt_func: Optional[Callable] = dc.field(default=None)
+    max_batch_size: Optional[int] = 64
+    inference_kwargs: dict = dc.field(default_factory=dict)
```

Comment:
What is this for? Passed to native model?

Reply:
Yes, different frameworks and API providers provide different parameter of `inference`/`generate` function.
Although they are mostly the same. 

But I left a parameter for users to adapt to them. If they don’t use it, they can ignore this parameter and the default parameters of the corresponding framework or API provider will be used.

Currently, we can pass in the parameters when generating text in two ways. This is one, and the other is through `db.predict("llm", max_tokens=100, **other_inference_kwargs)`

It will be explained in the documentation